Use the retrieved local K8s EndpointSlice in the CoreDNS plugin

tpantelis opened the pull request:
Fixes https://github.com/submariner-io/lighthouse/issues/1107

tpantelis committed:
Use the retrieved local K8s EndpointSlice in the CoreDNS plugin

Fixes https://github.com/submariner-io/lighthouse/issues/1107

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

vthapar commented:
https://github.com/submariner-io/lighthouse/pull/757#discussion_r873595715 gives some context to it. Not 100% sure if we need this check, in agent code these were added due to some issues with StatefulSets where some of the fields were empty or hostname/podname had a mismatch [don't recall exactly]. Also, issue was seen only in particular statefulset setup created by Raphael which, iirc, depended on how statefulset was defined. I do think we got it covered by E2E, so if E2e works, we should be fine.

So, while okay with this change, will need to keep an eye out for any SS issues that may crop up later.

vthapar reviewed (approved):
Tentatively approved as changes look fine, but would be good to make sure StatefulSets +Globalnet with local service works fine.

tpantelis commented:
> https://github.com/submariner-io/lighthouse/pull/757#discussion_r873595715 gives some context to it. Not 100% sure if we need this check, in agent code these were added due to some issues with StatefulSets where some of the fields were empty or hostname/podname had a mismatch [don't recall exactly].

So it seems there's a difference between how we derive our `EndpointSlice` addresses from the `Endpoints` and how K8s does it wrt the `Hostname` field. Notice [here](https://github.com/submariner-io/lighthouse/blob/6360c8a6f93db5cb6e46d180d13031380f980618/pkg/agent/controller/endpoint.go#L298) that we set `Hostname` from the `TargetRef` if `Hostname` isn't specifically set. I'm not clear on why we do this (shouldn't we just set the fields as is from the `EndpointAddress`?). I suspect K8s doesn't do this so `Hostname` is empty in this case, hence the mismatch. 

I'm not sure of the ramifications if this discrepancy but to preserve the behavior we could mimic our agent behavior here when replacing our `EndpointSlice` with K8s's.

aswinsuryan commented:
Is this going to break https://github.com/submariner-io/lighthouse/issues/993?

tpantelis commented:
> Is this going to break https://github.com/submariner-io/lighthouse/issues/993?

In what way? That PR was related to non-headless services.

tpantelis committed:
Determine host name for EndpointSlice addresses in plugin

On the agent side, when constructing an EndpointSlice, we set the
Hostname field from the TargetRef if Hostname isn't specifically set.
For consistency, we should mimic this behavior in the plugin for local
EndpointSlices when we elide the LH-specific EndpointSlice for globalnet.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented on coredns/resolver/endpoint_slice.go:
```diff
@@ -153,7 +154,17 @@ func (i *Interface) putHeadlessEndpointSlice(key, clusterID string, endpointSlic
+		switch {
+		case endpoint.Hostname != nil && *endpoint.Hostname != "":
+			hostname = *endpoint.Hostname
+		case endpoint.TargetRef != nil && strings.ToLower((*endpoint.TargetRef).Kind) == "pod":
```
@vthapar @aswinsuryan I assume we only want to interpret the TargetRef as a host name if it's a Pod?

vthapar replied:
Yes

tpantelis commented:
> I'm not sure of the ramifications if this discrepancy but to preserve the behavior we could mimic our agent behavior here when replacing our `EndpointSlice` with K8s's.

Pushed commit https://github.com/submariner-io/lighthouse/pull/1272/commits/d6beaad9c441453b4da2350b79826826e650a26b

vthapar commented:
> > [#757 (comment)](https://github.com/submariner-io/lighthouse/pull/757#discussion_r873595715) gives some context to it. Not 100% sure if we need this check, in agent code these were added due to some issues with StatefulSets where some of the fields were empty or hostname/podname had a mismatch [don't recall exactly].
> 
> So it seems there's a difference between how we derive our `EndpointSlice` addresses from the `Endpoints` and how K8s does it wrt the `Hostname` field. Notice [here](https://github.com/submariner-io/lighthouse/blob/6360c8a6f93db5cb6e46d180d13031380f980618/pkg/agent/controller/endpoint.go#L298) that we set `Hostname` from the `TargetRef` if `Hostname` isn't specifically set. I'm not clear on why we do this (shouldn't we just set the fields as is from the `EndpointAddress`?). I suspect K8s doesn't do this so `Hostname` is empty in this case, hence the mismatch.
> 
> I'm not sure of the ramifications if this discrepancy but to preserve the behavior we could mimic our agent behavior here when replacing our `EndpointSlice` with K8s's.

When we create and sync Endpointslice from agent, we don't add TargetRef field to it. This is to reduce the size of objects being synced for scale considerations. That is why we move information from TargetRef into hostname for Coredns plugin to consume on remote clusters. Similarly, in plugin we need to get hostname from TargetRef if hostname is empty as during testing, esp with StatefulSets I'd encountered situations where hostname was empty.

aswinsuryan commented:
> > Is this going to break #993?
> 
> In what way? That PR was related to non-headless services.

Oh ok I think I go it wrong, thanks for clarifying.

tpantelis committed:
Merge branch 'devel' into use_k8s_eps
